File: Library_Python/SearchEngine.py
from Media import Media
from Books import Books
from Film import Film
from Periodic import Periodic
from Video import Video
class SearchEngine:
    def __init__(self):
        self.cardCatalogue = []
        self.initialize_book()
        self.initialize_film()
        self.initialize_periodic()
        self.initialize_video()
    def initialize_book(self):
        with open("book.txt","r") as f:
            line = f.readline()
            while line!="":
                
                word = line.split("|")
                if (len(word)!= 10):
                    break
                bk = Books(word[0],word[1],word[2],word[3],word[4],word[5],word[6],word[7],word[8],word[9])
                self.cardCatalogue.append(bk)
                line = f.readline()

    def initialize_film(self):
        with open ("film.txt","r") as file_film:
            line = file_film.readline()
            while line!="":
                
                wordF = line.split("|")
                if (len(wordF )!= 6):
                    break
                fl = Film(wordF[0],wordF[1],wordF[2],wordF[3],wordF[4],wordF[5])
                self.cardCatalogue.append(fl)
                line= file_film.readline()
                
    def initialize_periodic(self):
        with open ("periodic.txt","r") as file_periodic:
            line = file_periodic.readline()
            while line!="":
                
                wordP = line.split("|")
                if(len(wordP)!= 12):
                    break
                pd = Periodic(wordP[0],wordP[1],wordP[2],wordP[3],wordP[4],wordP[5],wordP[6],wordP[7],wordP[8],wordP[9],wordP[10],wordP[11])
                self.cardCatalogue.append(pd)
                line = file_periodic.readline()
	
    def initialize_video(self):
        with open ("video.txt","r") as file_video:
            line = file_video.readline()
            while line!="":
               
                wordV = line.split("|")
                if (len(wordV)!=8):
                    break
                vd  = Video(wordV[0],wordV[1],wordV[2],wordV[3],wordV[4],wordV[5],wordV[6],wordV[7])
                self.cardCatalogue.append(vd)
                line = file_video.readline()
        
    def length(self):
        return len(self.cardCatalogue)
    def search_by_title(self,title):
        #search_byTitle= [];
        answer = False
        results = []
        sizeCC = self.length()
        for i in range (sizeCC):
            answer = self.cardCatalogue[i].compare_title(title)
            if (answer == True):
                results.append(self.cardCatalogue[i])
        return results;
    def search_by_call_number(self,callNumber):
        #search_callNumber= [];
        answer = False
        results= []
        sizeCC = self.length()
        for i in range (sizeCC):
            answer = self.cardCatalogue[i].compare_call_number(callNumber)
            #print (answer)
            if (answer == True):
                results.append(self.cardCatalogue[i])
        return results;
    def search_by_subjects (self, subjects):
        answer = False
        results = []
        sizeCC = self.length()
        for i in range (sizeCC):
            answer = self.cardCatalogue[i].compare_subjects(subjects)
            if (answer == True):
                results.append(self.cardCatalogue[i])
        return results;
    def search_by_other(self,other):
        answer = False
        results = []
        sizeCC = self.length();
        for i in range (sizeCC):
            answer = self.cardCatalogue[i].compare_other(other)
            if (answer == True):
                results.append(self.cardCatalogue[i])
        return results;

File: 511_library/SearchEngine.cpp
#include "SearchEngine.h"
#include <vector>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
//#include "Media.h"
#include <conio.h>


using namespace std;

SearchEngine::SearchEngine()
{
    initialize_books();
	initialize_film();
	initialize_Video();
	initialize_Periodic();
	
}

SearchEngine::~SearchEngine()
{
	for (int i=0; i<CardCatalog.size();i++)
	{
		delete CardCatalog[i];
	}
}

void SearchEngine::initialize_books()
{
	
	ifstream inf ("book.txt");
	string cn;
	string ti;
	string su;
	string au;
	string ds;
	string pb;
	string ct;
	string yr;
	string sr;
	string nts;
	//string

    //cout<<"Before while"<<endl;
    while (getline(inf, cn, '|'))
	{
		//getline(inf, cn, '|');
		getline(inf, ti, '|');
		getline(inf, su, '|');
		getline(inf, au, '|');
		getline(inf, ds, '|');
		getline(inf, pb, '|');
		getline(inf, ct, '|');
		getline(inf, yr, '|');
		getline(inf, sr, '|');
		getline(inf, nts, '\n');


		//cout << cn << ", " << ti <<", " << su <<", " << au <<", " << ds <<", " <<
		//pb <<  ", " << ct << ", " << yr << ", " <<sr << ", " <<nts << ", "<<endl;
        
        Books* temp = new Books (cn,ti,su,au,ds,pb,ct,yr,sr,nts);
        
        
        CardCatalog.push_back(temp);
       // cout<<"Printing things"<<endl;

	}
	inf.close();
            //display();
}

void SearchEngine::initialize_film()
{
	ifstream inf ("film.txt");
	string cn;
	string ti;
	string su;
	string di;
	string nts;
	string yr;

	while (getline(inf, cn, '|'))
	{
		//getline(inf, cn, '|');
		getline(inf, ti, '|');
		getline(inf, su, '|');
		getline(inf, di, '|');
		getline(inf, nts, '|');
		getline(inf, yr, '\n');


		Film* temp = new Film (cn,ti,su,di,nts,yr);
		CardCatalog.push_back(temp);
	}
	inf.close();

}


void SearchEngine::initialize_Periodic()
{
	ifstream inf ("periodic.txt");
	string cn;
	string ti;
	string su;
	string au;
	string de;
	string pu;
	string ph;
	string sr;
	string nts;
	string rt;
	string ofti;
	string goNu;

	while (getline(inf, cn, '|'))
	{
		//getline(inf, cn, '|');
		getline(inf, ti, '|');
		getline(inf, su, '|');
		getline(inf, au, '|');
		getline(inf, de, '|');
		getline(inf, pu, '|');
		getline(inf, ph, '|');
		getline(inf, sr, '|');
		getline(inf, nts, '|');
		getline(inf, rt, '|');
		getline(inf, ofti, '|');
		getline(inf, goNu, '\n');

		Periodic* temp = new Periodic(cn,ti,su,au,de,pu,ph,sr,nts,rt,ofti,goNu);
		CardCatalog.push_back(temp);
	}

	inf.close();
}

void SearchEngine::initialize_Video()
{
	ifstream inf("video.txt");
	string cn;
	string ti;
	string sb;
	string ds;
	string di;
	string nts;
	string sr;
	string la;

	while (getline(inf, cn, '|'))
	{
		
		getline(inf, ti, '|');
		getline(inf, sb, '|');
		getline(inf, ds, '|');
		getline(inf, di, '|');
		getline(inf, nts, '|');
		getline(inf, sr, '|');
		getline(inf, la, '\n');

		Video* temp = new Video(cn,ti,sb,ds,di,nts,sr,la);

		CardCatalog.push_back(temp);
	
	
	}
	inf.close();
}




/*void SearchEngine::display()
{
    for(int i = 0; i < (int)CardCatalog.size(); i++)
    {
        cout<<"Call number: "<<CardCatalog[i]->call_number<<endl;
        cout<<"Author: "<<CardCatalog[i]->title<<endl;
    }
}*/


vector<Media*> SearchEngine::search_by_title(const string& ss) const
{
	bool answer;
	std::vector <Media*> results;
	for (unsigned int i=0; i < CardCatalog.size(); i++)
	{
		answer = CardCatalog[i]->compare_title(ss);
		if (answer == true)
		{
			results.push_back(CardCatalog[i]);  
		}

	}
	return results;
}


vector<Media*> SearchEngine::search_by_call_number(const string& ss1) const
{
	bool answer;
	std::vector <Media*> results;
	for (unsigned int i=0; i < CardCatalog.size(); i++)
	{
		answer = CardCatalog[i]->compare_call_number(ss1);
		if (answer == true)
		{
			results.push_back(CardCatalog[i]);  
		}

	}
	return results;
}

vector< Media*> SearchEngine::search_by_subjects(const string& ss2) const
{
	bool answer;
	std::vector <Media*> results;
	for (unsigned int i=0; i < CardCatalog.size(); i++)
	{
		answer = CardCatalog[i]->compare_subjects(ss2);
		if (answer == true)
		{
			results.push_back(CardCatalog[i]);  
		}

	}
	return results;
}


vector< Media*> SearchEngine::search_by_others(const string& ss3) const
{
	bool answer ;
	std::vector <Media*> results;
	for (unsigned int i=0 ; i < CardCatalog.size();i++)
	{
		answer = CardCatalog[i]->compare_other(ss3);

		if(answer == true)
		{
			results.push_back(CardCatalog[i]);
		}

	}
	return results;

}